runtime/filetype.vim

```diff
@@ -1292,7 +1292,9 @@ au BufNewFile,BufRead *.jsonnet,*.libsonnet	setf jsonnet
 au BufNewFile,BufRead *.jl			setf julia
 
 " Just
-au BufNewFile,BufRead [jJ]ustfile,.justfile,*.just setf just
+au BufNewFile,BufRead \c{,*.}justfile,\c*.just setf just
+au BufNewFile,BufRead *
+	\ if getline(1) =~# '\v^#!/%(\w|[-/])*/%(env%(\s+-S)?\s+)?just\A' | setf just | endif
```

Comment:
I think this part belongs into `DetectFromHashBang()` in `autoload/dist/script.vim`